lib/charms/opensearch/v0/opensearch_relation_provider.py

```diff
@@ -311,16 +333,22 @@ def create_opensearch_users(
                 role_name=username,
                 permissions=self.get_extra_user_role_permissions(extra_user_roles, index),
             )
-            roles = [username]
-            self.user_manager.create_user(username, roles, hashed_pwd)
+            roles[rel_id].extend([username] + extra_user_roles.split(","))
+
+            self.user_manager.create_user(username, [username], hashed_pwd)
             self.user_manager.patch_user(
                 username,
-                [{"op": "replace", "path": "/opendistro_security_roles", "value": roles}],
+                [{"op": "replace", "path": "/opendistro_security_roles", "value": [username]}],
             )
+            users[rel_id].append(username)
         except OpenSearchUserMgmtError as err:
             logger.error(err)
             raise
 
+        if rel_id:
+            self.charm.peers_data.put(Scope.APP, ClientUsersDict, json.dumps(users))
+            self.charm.peers_data.put(Scope.APP, ClientRolesDict, json.dumps(roles))
```

Comment:
same comment about merging these 2 into a single dict - if necessary to store roles at all.

Reply:
Reworked, we only have a single list per relation now.